index.js

```diff
@@ -1,116 +1,238 @@
+  }
+
+  this.errors = errors;
 }
 
-IsoFactory.prototype.next = function(timestamp) {
-	if (typeof this.config.interval !== 'object') {
-		return NaN;
-	}
-	
-	var ret = moment(timestamp);
-	
-	if (ret.isValid() === false || ret > this.config.end_date) {
-		return NaN;
-	}
-	
-	ret.add(this.config.interval.years, 'years');
-	ret.add(this.config.interval.months, 'months');
-	ret.add(this.config.interval.days, 'days');
-	ret.add(this.config.interval.hours, 'hours');
-	ret.add(this.config.interval.minutes, 'minutes');
-	ret.add(this.config.interval.seconds, 'seconds');
-	
-	if (ret > this.config.end_date) { 
-		return NaN;
-	}
-
-	return ret.unix();
+Navigator.prototype.step = function(timestamp, back) {
+  var next = gmt(timestamp)
+    , prev = gmt(timestamp)
+    , next_count = 1
+    , prev_count = 0;
+
+  if (back) {
+    prev.subtract(this.duration);
+  } else {
+    next.add(this.duration);
+  }
+
+  return {next: next,
+          next_count: next_count,
+          prev: prev,
+          prev_count: prev_count};
+};
+
+Navigator.prototype.until = function(timestamp) {
+  var unlimited = this.repeats===Infinity
+    , next = gmt(this.start_date)
+    , next_count = 0
+    , prev
+    , prev_count;
+
+  if (this.start_date < timestamp) {
+    while (timestamp > next && (unlimited || this.repeats > next_count)) {
+      next_count++;
+      prev = gmt(next);
+      next.add(this.duration);
+    }
+
+    if (timestamp > next) { // we hit our repeat count
+      prev = next;
+      next = undefined;
+    }
+
+    prev_count = prev===undefined ? undefined : next_count - 1;
+    next_count = next===undefined ? undefined : next_count;
+  }
+
+  return {next: next,
+          next_count: next_count,
+          prev: prev,
+          prev_count: prev_count};
 };
 
-IsoFactory.prototype.previous = function(timestamp) {
-	if (typeof this.config.interval !== 'object') {
-		return NaN;
-	}
-	
-	var ret = moment(timestamp);
-	
-	if (ret.isValid() === false || ret < this.config.start_date) {
-		return NaN;
-	}
-	
-	ret.subtract(this.config.interval.years, 'years');
-	ret.subtract(this.config.interval.months, 'months');
-	ret.subtract(this.config.interval.days, 'days');
-	ret.subtract(this.config.interval.hours, 'hours');
-	ret.subtract(this.config.interval.minutes, 'minutes');
-	ret.subtract(this.config.interval.seconds, 'seconds');
-	
-	if (ret < this.config.start_date) { 
-		return NaN;
-	}
-	
-	return ret.unix();
+Navigator.prototype.backuntil = function(timestamp) {
+  var unlimited = this.repeats===Infinity
+    , prev = gmt(this.end_date)
+    , prev_count = this.repeats
+    , next
+    , next_count;
+
+  if (timestamp < this.end_date) {
+    while (timestamp <= prev && (unlimited || prev_count > 0)) {
+      prev_count--;
+      next = gmt(prev);
+      prev.subtract(this.duration);
+    }
+
+    if (timestamp < prev) {
+      next = prev;
+      prev = undefined;
+    }
+
+    next_count = next===undefined ? undefined : prev_count + 1;
```

Comment:
can we change to `(typeof next === 'undefined') ? undefined : prev_count + 1;`, just padding and wrapping ternary in parens in a couple places